### ISSUE TITLE
Support for custom datasets

### ISSUE THREAD

**Issue by dblunk88** [lowfuel/progrockdiffusion#80]
Would there be a possibility to natively implement an option to include datasets that are not listed in the settings.json? Perhaps have the script scan a separate folder and include all datasets in said folder.

**Comment by lowfuel** [comment]
Yeah, this is doable! I should have something for you tomorrow or Tuesday. However, please note I'm going to be traveling later next week, so if I can't get this in early in the week it will have to wait until July 5th or so (don't want to check in a big change and then leave for 5 days...)

**Comment by lowfuel** [comment]
Oh - quickly just want to add, it can't be as simple as scanning a directory for the files, because there are a bunch of settings that go with a diffusion model. So what I've done instead is moved the definitions of the models into a separate JSON file. All you have to do is edit that, add the model you want, and then use the new model name in your settings.json

**Comment by dblunk88** [comment]
You're awesome! :)

Looking at adding the Open Clip models ( https://github.com/mlfoundations/open_clip/releases/tag/v0.2-weights )

Might also be worth checking out for official support. It uses LAION, which is a pretty good dataset

**Comment by lowfuel** [comment]
Ah! So those are CLIP models. The stuff I was talking about above was for diffusion models, which are a totally separate thing.
Diffusion models just require a few configuration parameters, so they are easy to load from a settings file as you suggested.

With CLIP models (and OpenCLIP), their code also needs to know about the model. So either CLIP or OpenClip would have to be modified to handle loading of unknown models from a config file. I'm afraid you'd have to suggest this feature to them, but it might be impossible if actual code is required to support a new model.

Currently, the only way to support the models you mentioned above would be to convert Prog Rock to use OpenClip. Which may well be possible, I've not looked into it though. It says its an open source implementation of CLIP, so maybe a drop in replacement? Let me poke around a bit.

**Comment by dblunk88** [comment]
Let me know! If all else fails, I could see if I can modify a fork to enable those models via CLIP

**Comment by dblunk88** [comment]
> Sounds interesting for sure. Keep me posted! I'm not opposed to the idea, but CLIP is so essential to PRD that I'd want to be really sure the benefits outweigh any cons. :)

Should be! LAION has had some great results 
https://laion.ai/projects/

**Comment by dblunk88** [comment]
Reporting back (finally home and can test):

OpenAI is easily implemented. The only changes needed are:
1. Remove clip and "import open_clip as clip" instead
2. The model names,
3. "loaded_model = clip.load" to "loaded_model = clip.load_openai_model"

That will give you current functionality with OpenCLIP & included support for the LAION models

**Comment by dblunk88** [comment]
> Excellent, thanks. I will try it out locally and review.

Let me know if you can find out the t_int issue. If not, I'll keep on poking.
Also, I ran into an issue with OpenCLIP when trying to import, if that's the case for you as well, then try this fix: https://github.com/mlfoundations/open_clip/commit/9078293a67501d9489b22b8ee5b6c10b061a6b54

**Comment by dblunk88** [comment]
so, CLIP handles the load slightly different. OpenCLIP requires a direct .eval() instead of pulling [0]

Now I'm running into a conversion issue, am getting a: RuntimeError: expected scalar type Float but found Half

**Comment by lowfuel** [comment]
I looked at a few other notebooks that support both CLIP and OpenClip, and they seem to do so using this:

https://github.com/dmarx/Multi-Modal-Comparators

So I'll look into that as well, next week.

**Comment by nikocraft** [comment]
hey @dblunk88 what are the benefits of you using these new CLIP models? What will you gain? Can rest of us that use progrockdiffusion benefit, and how? same question for @lowfuel I'm not sure what the gain is with 
![image](https://user-images.githubusercontent.com/443080/176520449-396fcf2d-9e50-46e7-9d2f-c6c32b9b0940.png)

Are there examples of what we'd be able to do if lowfuel supports OpenClip?

**Comment by lowfuel** [comment]
Hard to say without trying, really, but it does seem that newer models released by OpenAI seem to improve the quality of turning a text prompt into guidance for diffusion. From what I understand a lot of DallE2's improvement comes from a new CLIP model they haven't released yet (VitH-16). Of course we'll get any new models released by OpenAI no problem, however PRD needs to be updated for these (rather trivial, luckily). 

The original request was to support models that don't have direct support within PRD, without requiring code to be updated, but that's not currently possible with CLIP, so it doesn't make much sense to support it in PRD. PRD would need to switch to something like OpenClip (evidently) that can read new models without any code changes.

What might be good is if someone were to fork PRD and convert it to use OpenClip, so that we can see some results from such a change first before deciding if it's worth pursuing. This is what I plan to look into next week, but honestly if it's a big change I probably will put it on the back burner for now.

**Comment by dblunk88** [comment]
> hey @dblunk88 what are the benefits of you using these new CLIP models? What will you gain? Can rest of us that use progrockdiffusion benefit, and how? same question for @lowfuel I'm not sure what the gain is with ![image](https://user-images.githubusercontent.com/443080/176520449-396fcf2d-9e50-46e7-9d2f-c6c32b9b0940.png)
> 
> Are there examples of what we'd be able to do if lowfuel supports OpenClip? will our renders get better more accurate?
> 
> I'd be interested if we as community could organise and train some more diffusion models for specific needs and distribute these so everyone can benefit, so we can get better rendering by plugging in correct diffusion model, for stuff (humans, creatures, vechicles) that DD currently has problems rendering correcly with etc... If we do this we'll stay up to date with MidJourney and maybe even Dalle-E 2 eventually...

LAION has actually been working on bettering things that are currently lacking. According to the Discord, they also have an unreleased model that helps generate humans. That being said, it shouldn't be too hard to custom train a model, the hard part is getting the data.

**Comment by nikocraft** [comment]
@dblunk88 maybe if you create a fork and get it working with OpenClip while lowfuel is away, he could take a look at it and see what improvments it brings and merge it in OR we could have it on a separate branch for a while, while we test it and see how we like it's results and performance. What do you think about that? Do you have time to give it a shot and make it work with OpenClip? I have neither time nor python skills, id I did I'd gladly help with that part :)

**Comment by dblunk88** [comment]
> @dblunk88 maybe if you create a fork and get it working with OpenClip while lowfuel is away, he could take a look at it and see what improvments it brings and merge it in OR we could have it on a separate branch for a while, while we test it and see how we like it's results and performance. What do you think about that? Do you have time to give it a shot and make it work with OpenClip? I have neither time nor python skills, id I did I'd gladly help with that part :)

I could take a stab at it. Took a little dive into the script before, but still not 100% sure how it works.

**Comment by dblunk88** [comment]
Got it working, it is way easier than I thought it was going to be

**Comment by dblunk88** [comment]
@lowfuel for mmc:
`
def load_clip_model(model_name: Text):
    print(f'-- {model_name}')
    with track_model_vram(device, model_name):
        loader = REGISTRY.find(architecture='clip', publisher='mlfoundations', id=model_name)[0]
        model = loader.load()
        loaded_model = mmc_mock_clip.MockOpenaiClip(model)
    return loaded_model`

**Comment by lowfuel** [comment]
Interesting! I worry a little that it's too many models (paradox of choice and all that), so I'm really curious to see what kind of results come from them. Have you tried any yet?

**Comment by dblunk88** [comment]
> Interesting! I worry a little that it's too many models (paradox of choice and all that), so I'm really curious to see what kind of results come from them. Have you tried any yet?

yeah, have tried some mlfoundations ones, they work great. Others I am having problems with from the ones I've tested (facebookresearch)

in order as not to clutter up the settings file, you could put in an 'advanced_models' option and then separate those out to a text file in a directory with the option of enabling or disabling them through said file. That would give the possibility of enabling those datasets without having a sea of datasets in the settings file itself

**Comment by lowfuel** [comment]
Very cool. And do you also need to have CLIP and OpenClip installed in addition to MMC? Or does MMC include everything?

**Comment by dblunk88** [comment]
I'm not sure, I already had both installed. Btw: some of these models you'll need napm which need to be loaded differently ( https://github.com/dmarx/Multi-Modal-Comparators/blob/main/README.md ). Currently working through this

**Comment by dblunk88** [comment]
![Test_5_0](https://user-images.githubusercontent.com/39381389/178075541-428c2f27-bcf4-4bd0-ad86-516bbcde4695.png)
![Test_6_0](https://user-images.githubusercontent.com/39381389/178075553-69b6d611-c70a-4b81-9082-3141603ed67e.png)
![Test1_4_0](https://user-images.githubusercontent.com/39381389/178075557-8da2847c-50b3-459a-bfa8-198df1299b4b.png)

Made with a mix of LAION + OpenAI datasets on a 3090 TI. First run for each.

**Comment by lowfuel** [comment]
Very cool! I'm knocked out by COVID at the moment, but once I'm feeling better I'll look into this.
